src/workerd/util/state-machine.h

```diff
@@ -0,0 +1,3897 @@
+      } else {
+        return kj::none;
+      }
+    }
+
+    ++this->transitionLockCount;
+    KJ_DEFER(--this->transitionLockCount);
+    const auto& active = this->state.template get<ActiveState>();
+    if constexpr (std::is_void_v<decltype(func(kj::instance<const ActiveState&>()))>) {
+      func(active);
+      return true;
+    } else {
+      return func(active);
+    }
+  }
+
+  // Execute a function if active, or return a default value.
+  // LOCKS TRANSITIONS during callback execution.
+  template <typename Func, typename Default>
+  auto whenActiveOr(
+      Func&& func, Default&& defaultValue) -> decltype(func(kj::instance<ActiveState&>())) {
+    if (!isActive()) {
+      return kj::fwd<Default>(defaultValue);
+    }
+
+    auto lock = this->acquireTransitionLock();
+    auto& active = this->state.template get<ActiveState>();
+    return func(active);
+  }
+
+  // ---------------------------------------------------------------------------
+  // State Transitions with Semantics
+  // ---------------------------------------------------------------------------
+
+  // Close the resource (transition to closed state)
+  template <typename... Args>
+  ClosedState& close(Args&&... args) KJ_LIFETIMEBOUND {
+    this->requireUnlocked();
+    KJ_REQUIRE(!isTerminated(), "Resource is already terminated");
+    return this->state.template init<ClosedState>(kj::fwd<Args>(args)...);
+  }
+
+  // Error the resource (transition to error state)
+  template <typename... Args>
+  ErrorState& error(Args&&... args) KJ_LIFETIMEBOUND {
+    this->requireUnlocked();
+    KJ_REQUIRE(!isTerminated(), "Resource is already terminated");
+    return this->state.template init<ErrorState>(kj::fwd<Args>(args)...);
+  }
+
+  // Close even if already terminated (for cleanup scenarios).
+  //
+  // WARNING: Bypasses terminal state protection. See forceTransitionTo() docs
+  // for when this is appropriate vs. suspicious.
+  template <typename... Args>
+  ClosedState& forceClose(Args&&... args) KJ_LIFETIMEBOUND {
+    this->requireUnlocked();
+    return this->state.template init<ClosedState>(kj::fwd<Args>(args)...);
+  }
+
+  // Error even if already terminated (for cleanup scenarios).
+  //
+  // WARNING: Bypasses terminal state protection. See forceTransitionTo() docs
+  // for when this is appropriate vs. suspicious.
+  template <typename... Args>
+  ErrorState& forceError(Args&&... args) KJ_LIFETIMEBOUND {
+    this->requireUnlocked();
+    return this->state.template init<ErrorState>(kj::fwd<Args>(args)...);
+  }
+
+  // Generic transition (for backward compatibility)
+  template <typename S, typename... Args>
+  S& transitionTo(Args&&... args) KJ_LIFETIMEBOUND
+    requires(_::isOneOf<S, ActiveState, ClosedState, ErrorState>)
+  {
+    this->requireUnlocked();
+    return this->state.template init<S>(kj::fwd<Args>(args)...);
+  }
+};
+
+// =============================================================================
+// Validated State Machine
```

Comment:
Unused, and not mentioned in this header's doc comment under "available classes"